engine/access/ingestion2/results_forest_test.go

```diff
@@ -0,0 +1,1684 @@
+	blocks := blockGen.List(50, blockGen.WithParentHeader(s.initialSealedHeader))
+	for _, block := range blocks {
+		s.allBlocks[block.ID()] = block
+	}
+
+	sealedResults := s.executionFork(blocks, s.initialSealedResult.ID())
+	conflictingReceipts := make([]*flow.ExecutionReceipt, 0)
+	for i, result := range sealedResults {
+		sealedPipeline := s.createPipeline(result)
+		sealedPipeline.On("SetSealed").Return().Once()
+		sealedPipeline.On("GetState").Return(optimistic_sync.StatePending).Maybe()
+
+		// create conflicting forks of up to 3 results descending from the parent of the sealed result
+		forkBlocks := blocks[i:min(i+3, len(blocks))]
+		forkResults := s.executionFork(forkBlocks, result.PreviousResultID)
+
+		// preconfigure the pipelines and results so insertions happen concurrently
+		for _, r := range forkResults {
+			// pipelines should return pending until they are abandoned. they will all be abandoned
+			// on insertion since they are either siblings of the sealed results, or descend from an
+			// abandoned parent.
+			isAbandoned := atomic.NewBool(false)
+			forkPipeline := s.createPipeline(r)
+			forkPipeline.On("GetState").Return(func() optimistic_sync.State {
+				if isAbandoned.Load() {
+					return optimistic_sync.StateAbandoned
+				}
+				return optimistic_sync.StatePending
+			}).Maybe()
+			forkPipeline.On("Abandon").Return().Run(func(args mock.Arguments) {
+				isAbandoned.Store(true)
+			})
+
+			receipt := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+			conflictingReceipts = append(conflictingReceipts, receipt)
+		}
+	}
+
+	// concurrently insert sealed results with AddSealedResult and conflicting receipts with AddReceipt
+	// AddSealedResult must be called sequentially to ensure the forest invariant is maintained and
+	// the results are actually added to the forest.
+	// synchronize goroutine startup to maximize concurrency
+	synctest.Test(s.T(), func(t *testing.T) {
+		start := make(chan struct{})
+		for _, receipt := range conflictingReceipts {
+			go func() {
+				<-start
+				added, err := forest.AddReceipt(receipt, ingestion2.ResultForCertifiedBlock)
+				s.Require().NoError(err)
+				s.True(added)
+			}()
+		}
+
+		// wait until all goroutines are running, then release them all at once
+		synctest.Wait()
+		close(start)
+
+		for _, result := range sealedResults {
+			err := forest.AddSealedResult(result)
+			s.Require().NoError(err)
+		}
+
+		synctest.Wait()
+	})
+
+	// there should be one result for each of the sealed and conflicting results, plus the latest
+	// persisted sealed result. All sealed results should be marked as sealed, and conflicting results
+	// should have the correct receipt and be marked Abandoned.
+	expectedSize := uint(1 + len(sealedResults) + len(conflictingReceipts))
+	s.Equal(expectedSize, forest.Size())
+
+	for _, result := range sealedResults {
+		s.assertContainer(forest, result.ID(), ingestion2.ResultSealed)
+	}
```

Comment:
Lets explicitly verify that the latest sealed result is correct

```suggestion
	var sealedResult *flow.ExecutionResult
	for _, sealedResult = range sealedResults {
		s.assertContainer(forest, sealedResult.ID(), ingestion2.ResultSealed)
	} // `sealedResult` should now reference the _latest_ sealed result 
	expectedLatestSealedResultBlock := s.allBlocks[sealedResult.BlockID]
	s.RequireLatestSealedResult(forest, sealedResult, expectedLatestSealedResultBlock.View)
	
```